// @remotion/transitions/.rollup.cache/Users/jonathanburger/remotion/packages/transitions/dist/presentations/clock-wipe.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { translatePath } from '@remotion/paths';
import { makePie } from '@remotion/shapes';
import { useMemo, useState } from 'react';
import { AbsoluteFill, random } from 'remotion';
const ClockWipePresentation = ({ children, presentationDirection, presentationProgress, passedProps }) => {
    const finishedRadius = Math.sqrt(passedProps.width ** 2 + passedProps.height ** 2) / 2;
    const { path } = makePie({
        radius: finishedRadius,
        progress: presentationProgress,
    });
    const translatedPath = translatePath(path, -(finishedRadius * 2 - passedProps.width) / 2, -(finishedRadius * 2 - passedProps.height) / 2);
    const [clipId] = useState(() => String(random(null)));
    const style = useMemo(() => {
        return {
            width: '100%',
            height: '100%',
            clipPath: presentationDirection === 'exiting' ? undefined : `url(#${clipId})`,
            ...(presentationDirection === 'entering'
                ? passedProps.innerEnterStyle
                : passedProps.innerExitStyle),
        };
    }, [
        clipId,
        passedProps.innerEnterStyle,
        passedProps.innerExitStyle,
        presentationDirection,
    ]);
    const outerStyle = useMemo(() => {
        return presentationDirection === 'entering'
            ? passedProps.outerEnterStyle
            : passedProps.outerExitStyle;
    }, [
        passedProps.outerEnterStyle,
        passedProps.outerExitStyle,
        presentationDirection,
    ]);
    return (_jsxs(AbsoluteFill, { style: outerStyle, children: [_jsx(AbsoluteFill, { style: style, children: children }), presentationDirection === 'exiting' ? null : (_jsx(AbsoluteFill, { children: _jsx("svg", { children: _jsx("defs", { children: _jsx("clipPath", { id: clipId, children: _jsx("path", { d: translatedPath, fill: "black" }) }) }) }) }))] }));
};
export const clockWipe = (props) => {
    return { component: ClockWipePresentation, props: props !== null && props !== void 0 ? props : {} };
};
