// @remotion/transitions/.rollup.cache/Users/jonathanburger/remotion/packages/transitions/dist/presentations/flip.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useMemo } from 'react';
import { AbsoluteFill, interpolate } from 'remotion';
const Flip = ({ children, presentationDirection, presentationProgress, passedProps: { direction = 'from-left', perspective = 1000, innerEnterStyle, innerExitStyle, outerEnterStyle, outerExitStyle, }, }) => {
    const style = useMemo(() => {
        const startRotationEntering = direction === 'from-right' || direction === 'from-top' ? 180 : -180;
        const endRotationEntering = direction === 'from-right' || direction === 'from-top' ? -180 : 180;
        const rotation = presentationDirection === 'entering'
            ? interpolate(presentationProgress, [0, 1], [startRotationEntering, 0])
            : interpolate(presentationProgress, [0, 1], [0, endRotationEntering]);
        const rotateProperty = direction === 'from-top' || direction === 'from-bottom'
            ? 'rotateX'
            : 'rotateY';
        return {
            width: '100%',
            height: '100%',
            transform: `${rotateProperty}(${rotation}deg)`,
            backfaceVisibility: 'hidden',
            WebkitBackfaceVisibility: 'hidden',
            ...(presentationDirection === 'entering'
                ? innerEnterStyle
                : innerExitStyle),
        };
    }, [
        direction,
        innerEnterStyle,
        innerExitStyle,
        presentationDirection,
        presentationProgress,
    ]);
    const outer = useMemo(() => {
        return {
            perspective,
            // Make children also their backface hidden
            transformStyle: 'preserve-3d',
            ...(presentationDirection === 'entering'
                ? outerEnterStyle
                : outerExitStyle),
        };
    }, [outerEnterStyle, outerExitStyle, perspective, presentationDirection]);
    return (_jsx(AbsoluteFill, { style: outer, children: _jsx(AbsoluteFill, { style: style, children: children }) }));
};
export const flip = (props) => {
    return { component: Flip, props: props !== null && props !== void 0 ? props : {} };
};
