// @remotion/transitions/.rollup.cache/Users/jonathanburger/remotion/packages/transitions/dist/timings/spring-timing.js
import { measureSpring, spring } from 'remotion';
const springWithInvalidArgumentRejection = (args) => {
    if (args.to || args.from) {
        throw new Error('to / from values are not supported by springWithRoundUpIfThreshold');
    }
    return spring(args);
};
export const springTiming = (options = {}) => {
    return {
        getDurationInFrames: ({ fps }) => {
            if (options.durationInFrames) {
                return options.durationInFrames;
            }
            return measureSpring({
                config: options.config,
                threshold: options.durationRestThreshold,
                fps,
            });
        },
        getProgress: ({ fps, frame }) => {
            return springWithInvalidArgumentRejection({
                fps,
                frame,
                config: options.config,
                durationInFrames: options.durationInFrames,
                durationRestThreshold: options.durationRestThreshold,
            });
        },
    };
};
